pandas/tests/typing/valid/test_pandas.py

```diff
@@ -0,0 +1,170 @@
+        {
+            "a": pd.Series([1, 2, 3]),
+            "b": pd.Series([4, 5, 6]),
+        },  # type:ignore[arg-type]
+        axis=1,
+    )
+
+    # error: Argument 1 to "concat" has incompatible type "Dict[str, DataFrame]";
+    # expected "Union[Iterable[DataFrame], Mapping[Hashable, DataFrame]]"
+    rdf1: pd.DataFrame = pd.concat({"a": df, "b": df2})  # type:ignore[arg-type]
+    # error: Argument 1 to "concat" has incompatible type "Dict[int, DataFrame]";
+    # expected "Union[Iterable[DataFrame], Mapping[Hashable, DataFrame]]"
+    rdf2: pd.DataFrame = pd.concat({1: df, 2: df2})  # type:ignore[arg-type]
+    # error: Argument 1 to "concat" has incompatible type "Dict[Optional[int],
+    # DataFrame]"; expected "Union[Iterable[DataFrame], Mapping[Hashable, DataFrame]]"
+    rdf3: pd.DataFrame = pd.concat({1: df, None: df2})  # type:ignore[arg-type]
+
+    rdf4: pd.DataFrame = pd.concat(list(map(lambda x: s2, ["some_value", 3])), axis=1)
+
+
+def test_types_json_normalize() -> None:
+    data1: List[Dict[str, Any]] = [
+        {"id": 1, "name": {"first": "Coleen", "last": "Volk"}},
+        {"name": {"given": "More", "family": "Regner"}},
+        {"id": 2, "name": "Faye Raker"},
+    ]
+    df1: pd.DataFrame = pd.json_normalize(data=data1)
+    df2: pd.DataFrame = pd.json_normalize(data=data1, max_level=0, sep=";")
+    df3: pd.DataFrame = pd.json_normalize(
+        data=data1, meta_prefix="id", record_prefix="name", errors="raise"
+    )
+    df4: pd.DataFrame = pd.json_normalize(data=data1, record_path=None, meta="id")
+    data2: Dict[str, Any] = {"name": {"given": "More", "family": "Regner"}}
+    df5: pd.DataFrame = pd.json_normalize(data=data2)
+
+
+def test_types_read_csv() -> None:
+    df = pd.DataFrame(data={"col1": [1, 2], "col2": [3, 4]})
+    # error: Incompatible types in assignment (expression has type "Optional[str]",
+    # variable has type "str")
+    csv_df: str = df.to_csv()  # type: ignore[assignment]
+
+    with tm.ensure_clean() as path:
+        df.to_csv(path)
+        df2: pd.DataFrame = pd.read_csv(path)
+        df3: pd.DataFrame = pd.read_csv(path, sep="a")
+        df4: pd.DataFrame = pd.read_csv(
+            path,
+            header=None,
+            mangle_dupe_cols=True,
+            keep_default_na=False,
+        )
+        df5: pd.DataFrame = pd.read_csv(
+            path, engine="python", true_values=[0, 1, 3], na_filter=False
+        )
+        df6: pd.DataFrame = pd.read_csv(
+            path,
+            skiprows=lambda x: x in [0, 2],
+            skip_blank_lines=True,
+            dayfirst=False,
+        )
+        df7: pd.DataFrame = pd.read_csv(path, nrows=2)
+        tfr1: TextFileReader = pd.read_csv(path, nrows=2, iterator=True, chunksize=3)
+        tfr2: TextFileReader = pd.read_csv(path, nrows=2, chunksize=1)
+        tfr3: TextFileReader = pd.read_csv(path, nrows=2, iterator=False, chunksize=1)
+        tfr4: TextFileReader = pd.read_csv(path, nrows=2, iterator=True)
```

Comment:
The TextFileReader tests create ResourceWarnings. I think the best is to have an empty `with` block after each line to close the internal file handles:

```py
tfr1: TextFileReader = pd.read_csv(path, nrows=2, iterator=True, chunksize=3)
with tfr1:
    pass
```

Reply:
(This creates failures on the Windows runs: cannot delete files that still have open file handles to it.)

Reply:
Tried to close it explicitly.

Thanks for looking at those issues. I usually push late and check the results the next day.